File: BilisPera/BilisPera/Utils/ADTool.m
//
//  ADTool.m
//  BilisPera
//
//  Created by BHJ on 2025/8/6.
//

#import "ADTool.h"
#import "KeychainWrapper.h"

NS_ASSUME_NONNULL_BEGIN

@implementation ADTool

+ (instancetype)shared {
    static ADTool *_shared = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _shared = [[ADTool alloc] init];
    });
    return _shared;
}

- (instancetype)init {
    if (self = [super init]) {
        _idfvString = [[[UIDevice currentDevice] identifierForVendor] UUIDString] ?: @"";
        _idfaString = @"";
        _trackCount = 0;
    }
    return self;
}

- (void)registerIDFAAndTrack {
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        __weak typeof(self) weakSelf = self;
        [self registerIDFAWithCompletion:^(BOOL success) {
            if (weakSelf.trackCount < 2) {
                [[TrackTools shared] trackForGoogleMarket];
                weakSelf.trackCount += 1;
            }
        }];
    });
}

- (void)registerIDFAWithCompletion:(void(^)(BOOL success))completion {
    if (@available(iOS 14, *)) {
        [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {
            dispatch_async(dispatch_get_main_queue(), ^{
                switch (status) {
                    case ATTrackingManagerAuthorizationStatusAuthorized: {
                        NSString *idfa = [[[ASIdentifierManager sharedManager] advertisingIdentifier] UUIDString];
                        ADTool.shared.idfaString = idfa;
                        NSData *idfvData = [self.idfvString dataUsingEncoding:NSUTF8StringEncoding];
                        if (idfvData) {
                            [KeychainWrapper saveWithKey:@"IDFV" data:idfvData];
                        }
                        if (completion) completion(YES);
                        break;
                    }
                    case ATTrackingManagerAuthorizationStatusNotDetermined:
                        [self registerIDFAWithCompletion:completion];
                        break;
                    default:
                        if (completion) completion(NO);
                        break;
                }
            });
        }];
    } else {
        if ([[ASIdentifierManager sharedManager] isAdvertisingTrackingEnabled]) {
            NSString *idfa = [[[ASIdentifierManager sharedManager] advertisingIdentifier] UUIDString];
            self.idfaString = idfa;
            
            NSData *idfvData = [self.idfvString dataUsingEncoding:NSUTF8StringEncoding];
            if (idfvData) {
                [KeychainWrapper saveWithKey:@"IDFV" data:idfvData];
            }
            if (completion) completion(YES);
        } else {
            if (completion) completion(NO);
        }
    }
}

@end

NS_ASSUME_NONNULL_END

File: SkyLoan_phillipines/Utils/ADTool.swift
//
//  ADTool.swift
//  SkyLoan_phillipines
//
//  Created by BHJ on 2025/5/20.
//

import Foundation
import AdSupport
import AppTrackingTransparency
import UIKit

class ADTool {
    static let shared = ADTool()
    let idfvString: String = UIDevice.current.identifierForVendor?.uuidString ?? ""
    var idfaString: String = ""
    var trackCount: Int = 0
    
    func registerIDFAAndTrack(){
        Task{
            let delay = UInt64(1.5) * 1_000_000_000
            try await Task.sleep(nanoseconds: delay)
            guard await registerIDFA() else {return}
            guard self.trackCount < 2 else {return}
            TrackMananger.shared.trackGoogleMarket()
            self.trackCount += 1
        }
    }
    
    func registerIDFA() async -> Bool{
        let status =  await ATTrackingManager.requestTrackingAuthorization()
        switch status {
        case .authorized:
            // 用户已授权，可以访问IDFA
            let idfa = ASIdentifierManager.shared().advertisingIdentifier.uuidString
            ADTool.shared.idfaString = idfa
            if let idfvData = self.idfvString.data(using: .utf8){
                _ = KeychainWrapper.save(key: "IDFV", data: idfvData)
            }
            return true
        case .notDetermined:
            return await registerIDFA()
        case .denied, .restricted:
            return false
        @unknown default:
            return false
        }
    }
}
